tidy3d/plugins/mode/solver.py

```diff
@@ -508,10 +505,30 @@ def any_pec(eps_vec, threshold=0.9 * np.abs(pec_val)):
             vec_init = dnz * vec_init
 
         if enable_preconditioner:
-            precon = sp.diags(1 / mat.diagonal())
-            mat = mat * precon
+
+            def conditional_inverted_vec(eps_vec, threshold=1):
+                """Returns a diagonal sparse matrix whose i-th element in the diagonal
+                is |eps_i|^-1 if |eps_i|>threshold, and |eps_i| otherwise.
+                """
+                abs_vec = np.abs(eps_vec)
+                return sp.spdiags(np.where(abs_vec > threshold, 1.0 / abs_vec, abs_vec), [0], N, N)
```

Comment:
Very clever fix, hope it works! I took a look at your notes and it makes sense, but I do not understand the `conditional_inverted_vec` function here. You take 1 / eps if eps > 1 otherwise you take eps? How does that make sense?

Reply:
This is to make sure to scale down epsilon when it’s huge, but we don’t want to make the matrix ill conditioned if epsilon is close to 0

Reply:
And more importantly, unphysical?

Reply:
It doesn’t change the eigenvalue problem, as it’s just preconditioner.

Reply:
Ah ok that's the part I missed!

Reply:
Yeah, that’s in the last section of the notes